core/bench_test.go

```diff
@@ -0,0 +1,109 @@
+package core
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"chain/core/account"
+	"chain/core/asset"
+	"chain/core/coretest"
+	"chain/core/generator"
+	"chain/core/pin"
+	"chain/core/txbuilder"
+	"chain/database/pg/pgtest"
+	"chain/protocol/bc"
+	"chain/protocol/prottest"
+	"chain/testutil"
+)
+
+func BenchmarkBuildTx(b *testing.B) {
+	b.StopTimer()
+
+	cases := []struct {
+		numAssets, numIssuances int
+	}{
+		{1, 1}, {1, 10}, {1, 100},
+		{10, 1}, {10, 10}, {10, 100},
+		{100, 1}, {100, 10},
+	}
+
+	_, db := pgtest.NewDB(b, pgtest.SchemaPath)
+	ctx := context.Background()
+	pinStore := pin.NewStore(db)
+	chain := prottest.NewChain(b)
+	accounts := account.NewManager(db, chain, pinStore)
+	assets := asset.NewRegistry(db, chain, pinStore)
+	accountID := coretest.CreateAccount(ctx, b, accounts, "account", nil)
+	generator := generator.New(chain, nil, db)
+
+	var assetIDs []bc.AssetID
+	for _, c := range cases {
+		if c.numAssets > len(assetIDs) {
+			for i := len(assetIDs); i < c.numAssets; i++ {
+				assetID := coretest.CreateAsset(ctx, b, assets, nil, fmt.Sprintf("asset%d", i), nil)
+				assetIDs = append(assetIDs, assetID)
+			}
+		}
+	}
+
+	prepareActions := func(numAssets, numIssuances int) []txbuilder.Action {
+		var actions []txbuilder.Action
+		for i := 0; i < numAssets; i++ {
+			assetID := assetIDs[i]
+			for j := 0; j < numIssuances; j++ {
+				actions = append(actions, assets.NewIssueAction(bc.AssetAmount{AssetId: &assetID, Amount: 1}, nil))
+			}
+			actions = append(actions, accounts.NewControlAction(bc.AssetAmount{AssetId: &assetID, Amount: uint64(numIssuances)}, accountID, nil))
+		}
+		return actions
+	}
+
+	doBuild := func(actions []txbuilder.Action) *txbuilder.Template {
+		tpl, err := txbuilder.Build(ctx, nil, actions, time.Now().Add(time.Minute))
+		if err != nil {
+			b.Fatal(err)
+		}
+		return tpl
+	}
+
+	for _, c := range cases {
+		name := fmt.Sprintf("%d-asset--%d-issuance", c.numAssets, c.numIssuances)
+		b.Run(name+"--build", func(b *testing.B) {
+			actions := prepareActions(c.numAssets, c.numIssuances)
+			for i := 0; i < b.N; i++ {
+				b.StartTimer()
```

Comment:
Were you trying to exclude the time spent in prepareActions? A *testing.B starts out with its timer started so you'd have to call StopTimer before prepareActions.

Also there's no point in calling StopTimer at the end of the loop body because the only thing between that and the StartTimer at the top of the loop body is the branch to the beginning of the loop. 

It's more typical to see this sort of thing:

```go
b.Run(name+"--build", func(b *testing.B) {
	b.StopTimer()
	actions := prepareActions(c.numAssets, c.numIssuances)
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		doBuild(actions)
	}
})
```